soroban-env-host/src/host.rs

```diff
@@ -2771,10 +2771,45 @@ impl VmCallerEnv for Host {
         signature: BytesObject,
         recovery_id: U32Val,
     ) -> Result<BytesObject, HostError> {
+        #[cfg(not(feature = "next"))]
         let sig = self.secp256k1_signature_from_bytesobj_input(signature)?;
+        #[cfg(feature = "next")]
+        let sig = self.ecdsa_signature_from_bytesobj_input::<k256::Secp256k1>(signature)?;
         let rid = self.secp256k1_recovery_id_from_u32val(recovery_id)?;
         let hash = self.hash_from_bytesobj_input("msg_digest", msg_digest)?;
-        self.recover_key_ecdsa_secp256k1_internal(&hash, &sig, rid)
+        let rk = self.recover_key_ecdsa_secp256k1_internal(&hash, &sig, rid)?;
+        self.add_host_object(rk)
+    }
+
+    #[cfg(feature = "next")]
+    fn verify_sig_ecdsa_secp256r1(
+        &self,
+        _vmcaller: &mut VmCaller<Host>,
+        public_key: BytesObject,
+        msg_digest: BytesObject,
+        signature: BytesObject,
+    ) -> Result<Void, HostError> {
+        let pk = self.secp256r1_public_key_from_bytesobj_input(public_key)?;
+        let sig = self.ecdsa_signature_from_bytesobj_input::<p256::NistP256>(signature)?;
+        let msg_hash = self.hash_from_bytesobj_input("msg_digest", msg_digest)?;
+        let res = self.secp256r1_verify_signature(&pk, &msg_hash, &sig)?;
+        Ok(res.into())
+    }
+
+    #[cfg(not(feature = "next"))]
```

Comment:
This looks a bit suspicious to me; shouldn't this check be performed at runtime? p21 build will still run p20 for a while. Or is this supposed to be unreachable?

Reply:
Yeah, I might be a bit too paranoid with "next" gating here.
The new host function is runtime gated already (via `min_supported_protocol` gating). So we don't need the gate here, as well as most other helper functions. I'm going to remove the compile gate here to simplify this PR a bit (and making it a little easier to look at). 
I was thinking compile-time gating this new functions would essentially compile out new external dependencies (p256 crate etc.). But I don't think this is necessary (the code footprint is small enough and is entirely within our control).

Reply:
Unfortunately since the xdr "curr" and "next" has already differed on the new cost type, which is needed for the charge budget call in the new methods. The easiest way to gate away the difference is to do it here, and gating all the helper methods it calls. (Compared to gating the `charge_budget` call that uses the new xdr definition)
I think this is fine. The function is already gated by run-time protocol-gating, here just a bit redundant and it will be removed soon.

Reply:
Ok, that works. Just wanted to point out that we wouldn't have needed excessive gating if we merged the XDR changes to `curr` branch.